src/API/Data/UnitsRepository.cs

```diff
@@ -34,14 +35,103 @@ public class UnitsRepository : DataRepository
+            => await ExecuteDbPipeline("create a unit", db =>
+                TrySetRequestedParent(db, body)
+                .Bind(parent => TryCreateUnit(db, body)));
+
+        internal static async Task<Result<Unit, Error>> UpdateUnit(UnitRequest body, int unitId)
+        {
+            return await ExecuteDbPipeline($"update unit {unitId}", db =>
+                TrySetRequestedParent(db, body)
+                .Bind(_ => TryFindUnit(db, unitId))
+                .Bind(existing => TryUpdateUnit(db, existing, body))
+            );
+        }
+
+        internal static async Task<Result<bool, Error>> DeleteUnit(int unitId)
         {
-            var person = await db.Units
+            return await ExecuteDbPipeline($"delte unit {unitId}", db =>
+                TryFindUnit(db, unitId)
+                .Bind(unit => TryDeleteUnit(db, unit)));
+        }
+
+        private static async Task<Result<Unit,Error>> TryFindUnit (PeopleContext db, int id, bool findingParent = false)
+        {
+            var unit = await db.Units
                 .Include(u => u.Parent)
                 .SingleOrDefaultAsync(p => p.Id == id);
-            return person == null
-                ? Pipeline.NotFound("No unit found with that ID.")
-                : Pipeline.Success(person);
+            return unit == null
+                ? Pipeline.NotFound($"No {(findingParent ? "parent" : "")} unit found with ID ({id}).")
+                : Pipeline.Success(unit);
+        }
+
+        private static async Task<Result<UnitRequest,Error>> TrySetRequestedParent (PeopleContext db, UnitRequest body)
+        {
+            if(body.ParentId.HasValue && body.ParentId != 0)
+            {
+                return await TryFindUnit(db, (int)body.ParentId, true)
+                    .Tap(p => body.SetParent(p))
+                    .Bind(_ => Pipeline.Success(body));
+            }
+            else
+            {
+                return Pipeline.Success(body);
+            }
+        }
+
+        private static async Task<Result<Unit,Error>> TryCreateUnit (PeopleContext db, UnitRequest body)
+        {
+            var unit = new Unit(body.Name, body.Description, body.Url, body.Email, body.GetParent());
+
+            // add the unit
+            db.Units.Add(unit);
+            // TODO: We might need to force the validation of unit before EF saves it to the DB.
+            // save changes
+            await db.SaveChangesAsync();
+            return Pipeline.Success(unit);
+        }
+
+        private static async Task<Result<Unit, Error>> TryUpdateUnit(PeopleContext db, Unit existing, UnitRequest body)
+        {
+            existing.Name = body.Name;
+            existing.Description = body.Description;
+            existing.Url = body.Url;
+            existing.Email = body.Email;
+            existing.Parent = body.GetParent();
+
+            // TODO: Do we need to manually validate the model?  EF doesn't do that for free any more.
```

Comment:
The request body should be getting validated immediately after its deserialized. See `Api.Middleware.Request.DeserializeBody()`. EF should observe any property annotations. If that's _not_ happening, lets dig into it.

Reply:
I'll test this out more, but I know somewhere along my testing I wound up with Units inserted that didn't have `Name` which is a required field.

Reply:
It does validate the `UnitRequest` DTO, but does not validate the resulting `Unit` entity before committing it to the database.

If there's a gap in your code that maps the DTO to its entity - or you add a Data Annotation to the entity, but not the DTO - you could end up with malformed data written to the database.

It's some belts-and-bracers protection I miss from old Entity Framework.  I might try to piggyback this validation in when we start auditing all changes to entities.